engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/EmailValidator.java

```diff
@@ -69,29 +71,50 @@ public boolean isValid(CharSequence value, ConstraintValidatorContext context) {
 		}
 
 		// if we have a trailing dot in local or domain part we have an invalid email address.
-		// the regular expression match would take care of this, but IDN.toASCII drops trailing the trailing '.'
+		// the regular expression match would take care of this, but IDN.toASCII drops the trailing '.'
 		// (imo a bug in the implementation)
 		if ( emailParts[0].endsWith( "." ) || emailParts[1].endsWith( "." ) ) {
 			return false;
 		}
 
-		if ( !matchPart( emailParts[0], localPattern ) ) {
+		if ( !matchPart( emailParts[0], localPattern, MAX_LOCAL_PART_LENGTH ) ) {
 			return false;
 		}
 
-		return matchPart( emailParts[1], domainPattern );
+		return matchPart( emailParts[1], domainPattern, MAX_DOMAIN_PART_LENGTH );
 	}
 
-	private boolean matchPart(String part, Pattern pattern) {
+	private boolean matchPart(String part, Pattern pattern, int maxLength) {
+		String asciiString;
 		try {
-			part = IDN.toASCII( part );
+			asciiString = toAscii( part );
 		}
 		catch ( IllegalArgumentException e ) {
-			// occurs when the label is too long (>63, even though it should probably be 64 - see http://www.rfc-editor.org/errata_search.php?rfc=3696,
-			// practically that should not be a problem)
 			return false;
 		}
-		Matcher matcher = pattern.matcher( part );
+
+		if ( asciiString.length() > maxLength ) {
+			return false;
+		}
+
+		Matcher matcher = pattern.matcher( asciiString );
 		return matcher.matches();
 	}
+
+	private String toAscii(String unicodeString) throws IllegalArgumentException {
+		String asciiString = "";
+		int start = 0;
+		int end = unicodeString.length() <= 63 ? unicodeString.length() : 63;
+		while ( true ) {
```

Comment:
Looks alright, though I find the version from the [second answer here](http://stackoverflow.com/questions/3760152/split-string-to-equal-length-substrings-in-java) slightly more digestable.